src/bakery/Bakery.cpp

```diff
@@ -188,17 +188,143 @@ void Bakery::greedyWithDijkstra(Van& van) {
+}
+
+int Bakery::knapsackAllocation(Van &v, const vector<int>& values) {
+    vector<Client*> clients;
+    for (Client* client : this->clients)
+        if (!client->isAllocated()) clients.push_back(client);
+
+    vector<vector<int>> table(clients.size() + 1, vector<int>(v.getTotalBread() + 1));
+
+    for (int i = 0; i <= clients.size(); ++i)
+        for (int w = 0; w <= v.getTotalBread(); ++w) {
+            if (i == 0 || w == 0)
+                table[i][w] = 0;
+            else if (clients[i - 1]->getBreadQuantity() <= w)
+                table[i][w] = max(values[i - 1] + table[i - 1][w - clients[i - 1]->getBreadQuantity()],
+                                  table[i - 1][w]);
+            else
+                table[i][w] = table[i - 1][w];
+        }
+
+
+    int w = v.getTotalBread(), i = clients.size(), removed = 0;
+    while (i > 0) {
+        if (table[i][w] - table[i - 1][w - clients[i - 1]->getBreadQuantity()] == values[i - 1]) {
+            // Element i should be allocated
+            v.addClient(clients[i - 1]);
+            removed++;
+            w -= clients[i - 1]->getBreadQuantity();
+        }
+        i--;
+    }
+    return removed;
+}
+
+int Bakery::greedyAllocation(Van &v) {
+    int count = 0;
+    int capacity = v.getTotalBread();
+
+    while (true) {
+        int highestValue = -INF;
+        Client *chosen = NULL;
+
+        /*
+         * Value = 50000 + 10 * numBreads - 100 * Euclidean(start, pos)
+         * Summation(10 * abs(v.clients[i].time - time) - 100 * Euclidean(v.clients[i].pos, pos))
+         */
+        for (Client *client : clients) {
+            if (client->isAllocated()) continue;
+
+            int value = 50000 + 10 * client->getBreadQuantity() - 100 *
+                    startingVertex->getPosition().distance(client->getVertex()->getPosition());
+
+            for (Client *client2 : v.getClients())
+                value += abs((client2->getRealTime() - client->getRealTime()).toMinutes())
+                        - 100 * client2->getVertex()->getPosition().distance(client->getVertex()->getPosition());
+
+            if (value > highestValue && client->getBreadQuantity() <= capacity) {
+                highestValue = value;
+                chosen = client;
+            }
+        }
+        if (chosen == NULL) break;
+        v.addClient(chosen);
+        count++;
+        capacity -= chosen->getBreadQuantity();
+    }
+    return count;
+}
+
+void Bakery::allocateClientsToVans(bool useKnapsack, bool optimize) {
+    sort(vans.begin(), vans.end(), [](const Van& v1, const Van& v2) -> bool {
+        return v1.getTotalBread() > v2.getTotalBread();
+    });
+
+    int clientsAllocated = 0;
+
+    if (useKnapsack) {
+        // Value = breadQuantity / 10 + 1 (the client itself has value)
+        vector<int> values;
+        for (Van &v : vans) {
+            if (clientsAllocated == clients.size()) break;
+
+            for (Client *client : clients)
```

Comment:
Are the clients sorted by position? By latitude or something of the sort. Since that way each van would have the clients closer together

Reply:
That's an interesting idea but it also doesn't seem that easy. Even if clients have the same latitude, they can still be very far away. But I put a TODO note so we explore this later